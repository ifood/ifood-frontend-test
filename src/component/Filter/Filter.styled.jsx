import { Input } from 'semantic-ui-react';
import styled from 'styled-components';

const Search = styled.div`
  display: flex;
  margin-bottom: 20px;
`;

const ModalTriggerWrapper = styled.div`
  width: 55px;
  flex-shrink: 0;
  align-items: center;
  display: flex;
  justify-content: flex-end;
`;

const InputSearch = styled(Input)`
  flex-grow: 1;
`;

export { Search, ModalTriggerWrapper, InputSearch };
