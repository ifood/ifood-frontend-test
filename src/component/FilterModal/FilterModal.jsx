import React from 'react';
import PropTypes from 'prop-types';
import { Button, Modal } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';
import DynamicForm from '../DynamicForm';

function FilterModal({ open, filter, metadata, onSubmit, onClose }) {
  function handleClose() {
    if (onClose != null) {
      onClose();
    }
  }

  function handleSubmit(data) {
    if (onSubmit != null) {
      onSubmit(data);
    }
    handleClose();
  }

  return (
    <Modal size="tiny" onClose={handleClose} open={open} centered={false}>
      <Modal.Header>
        <FormattedMessage id="filter" />
      </Modal.Header>
      <Modal.Content>
        <DynamicForm
          id="filterForm"
          data={filter}
          metadata={metadata}
          onSubmit={handleSubmit}
        />
      </Modal.Content>
      <Modal.Actions>
        <Button id="modalFilter-cancel" color="black" onClick={handleClose}>
          <FormattedMessage id="cancel" />
        </Button>
        <Button type="submit" form="filterForm" color="red">
          <FormattedMessage id="filter.submit" />
        </Button>
      </Modal.Actions>
    </Modal>
  );
}

FilterModal.propTypes = {
  open: PropTypes.bool,
  filter: PropTypes.object,
  metadata: PropTypes.array,
  onSubmit: PropTypes.func,
  onClose: PropTypes.func,
};

export default FilterModal;
