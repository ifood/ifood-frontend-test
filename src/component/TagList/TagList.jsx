import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Tag from '../Tag';

const Tags = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

function TagList({ title, items, onDeleteTag }) {
  if (items == null || items.length === 0) {
    return null;
  }

  return (
    <>
      {title != null && <span>{title}:</span>}
      <Tags>
        {items.map((item) => {
          const { id, field, value } = item;
          return (
            <Tag
              key={id}
              id={id}
              field={field}
              value={value}
              onDelete={onDeleteTag}
            />
          );
        })}
      </Tags>
    </>
  );
}

TagList.propTypes = {
  title: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      field: PropTypes.string,
      value: PropTypes.string,
    })
  ),
  onDeleteTag: PropTypes.func,
};

export default TagList;
