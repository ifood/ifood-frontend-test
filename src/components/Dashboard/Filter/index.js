/* eslint-disable react/jsx-props-no-spreading */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FaSearch, FaFilter, FaAngleDown, FaAngleUp } from 'react-icons/fa';
import Select from 'react-select';
import DatePicker, { registerLocale } from 'react-datepicker';
import pt from 'date-fns/locale/pt';

import useSkipFirstRender from '../../../utils/useSkipFirstRender';

import { Container, Label } from './styles';
import 'react-datepicker/dist/react-datepicker.css';

registerLocale('pt', pt);

function Filter({
  playlists,
  setPlaylists,
  setFilters,
  getPlaylistsService,
  filtersLabelValue,
}) {
  const [localeOptions, setLocaleOptions] = useState([]);
  const [countryOptions, setCountryOptions] = useState([]);
  const [playlistsInitial, setPlaylistsInitial] = useState([]);
  const [values, setValues] = useState({});
  const [isShowingFilter, setIsShowingFilter] = useState(false);
  const [searchValueFocus, setSearchValueFocus] = useState(false);

  const handleSearchInputChange = (event) => {
    event.persist();

    const searchValue = event.target.value;

    if (searchValue.length) {
      const filteredItems = playlistsInitial.filter(({ name }) =>
        name.toLocaleLowerCase().includes(searchValue.toLocaleLowerCase()),
      );

      setValues((prevState) => ({ ...prevState, searchValue }));
      setPlaylists(filteredItems);
    } else {
      setPlaylists(playlistsInitial);
      setValues((prevState) => ({ ...prevState, searchValue: '' }));
    }
  };

  const handleInputChange = (name, value) => {
    setValues((prevState) => ({ ...prevState, [name]: value }));
  };

  const handleDateTimeChange = (value) => {
    setValues((prevState) => ({ ...prevState, date: value }));
  };

  useSkipFirstRender(() => {
    const { searchValue, locale, country, date, limit, offset } = values;

    setFilters({
      searchValue,
      locale: locale || 'pt_BR',
      country: country || 'BR',
      timestamp: (date && date.toISOString()) || new Date().toISOString(),
      limit: limit || 12,
      offset: offset || 0,
    });
    getPlaylistsService();
  }, [values]);

  useEffect(() => {
    if (!playlistsInitial.length) {
      setPlaylistsInitial(playlists);
    }

    if (filtersLabelValue) {
      let locales = filtersLabelValue.find((value) => value.id === 'locale');
      let countries = filtersLabelValue.find((value) => value.id === 'country');

      locales = locales.values.map((item) => ({ ...item, label: item.name }));
      countries = countries.values.map((item) => {
        let { value } = item;
        let label = item.name;

        if (value === 'en_US') {
          value = 'US';
          label = 'Estados Unidos';
        }

        return { value, label };
      });

      setLocaleOptions(locales);
      setCountryOptions(countries);
    }
  }, [playlists, playlistsInitial, filtersLabelValue]);

  return (
    <Container showFilter={isShowingFilter}>
      <div className="filter-wrapper">
        <div className="filter-content">
          <div className="filter-search-container">
            <Label focus={searchValueFocus}>
              <FaSearch size={16} className="fa fa-search" />
              <input
                name="searchValue"
                type="text"
                placeholder="Filtrar"
                onChange={handleSearchInputChange}
                onFocus={() => setSearchValueFocus(true)}
                onBlur={() => setSearchValueFocus(false)}
              />
            </Label>
            <button
              type="button"
              className="filter-button"
              onClick={() => setIsShowingFilter(!isShowingFilter)}
            >
              <FaFilter size={16} className="fa fa-filter" />
              {isShowingFilter ? (
                <FaAngleUp size={18} className="fa fa-angle-down" />
              ) : (
                <FaAngleDown size={18} className="fa fa-angle-down" />
              )}
            </button>
          </div>
          <div className="filter-items">
            <Select
              name="locale"
              className="react-select"
              classNamePrefix="select"
              placeholder="Localidade"
              isSearchable={false}
              options={localeOptions}
              onChange={(value) => handleInputChange('locale', value.value)}
            />
            <Select
              name="country"
              className="react-select"
              classNamePrefix="select"
              placeholder="País"
              isSearchable={false}
              options={countryOptions}
              onChange={(value) => handleInputChange('country', value.value)}
            />
            <DatePicker
              showTimeInput
              locale="pt"
              timeInputLabel="Horário:"
              timeFormat="p"
              dateFormat="Pp"
              placeholderText="Dia e horário"
              className="default-input"
              selected={values.date}
              onChange={(date) => handleDateTimeChange(date)}
            />
            <input
              name="limit"
              type="number"
              placeholder="Itens por página"
              className="default-input"
              onChange={(e) =>
                e.target.value <= 60 &&
                handleInputChange(e.target.name, e.target.value)
              }
              value={values.limit}
            />
            <input
              name="offset"
              type="number"
              placeholder="Página"
              className="default-input"
              onChange={(e) => handleInputChange(e.target.name, e.target.value)}
              value={values.offset}
            />
          </div>
        </div>
      </div>
    </Container>
  );
}

Filter.defaultProps = {
  filtersLabelValue: [],
};

Filter.propTypes = {
  filtersLabelValue: PropTypes.arrayOf(PropTypes.any),
  playlists: PropTypes.arrayOf(PropTypes.any).isRequired,
  setPlaylists: PropTypes.func.isRequired,
  setFilters: PropTypes.func.isRequired,
  getPlaylistsService: PropTypes.func.isRequired,
};

export default Filter;
