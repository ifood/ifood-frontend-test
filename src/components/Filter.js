// import { useEffect, useState } from 'react'
import usePlaylists from '../contexts/PlaylistContext'
import FieldsFilter from './FieldsFilter'

const Filter = () => {
  const { filterFields, loadingFilterFields } = usePlaylists()

  return (
    <>
      {loadingFilterFields ? (
        <p>Loading Filter</p>
      ) : (
        <div>
          {filterFields.map((filter) => (
            <div key={filter.id}>
              <h3>{filter.name}</h3>
              <FieldsFilter filter={filter} />
              {/* <ul>
                {filter.values &&
                  filter.values.map((option) => <p>{option.name}</p>)}
              </ul> */}
            </div>
          ))}
        </div>
      )}
    </>
  )
}

export default Filter
