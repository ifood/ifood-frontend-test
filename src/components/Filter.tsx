import React, { PureComponent } from 'react';

interface IProps {
  onSignOut: () => void;
}

export default class Filter extends PureComponent<IProps> {
  public render() {
    return (
      <div>
        <h1>Hi, I'm the filter component</h1>
        <input type='text' />
        <select />
        <button onClick={this.props.onSignOut}>Sign out</button>
      </div>
    );
  }
}
