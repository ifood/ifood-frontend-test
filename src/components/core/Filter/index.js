import React from "react";
import PropTypes from "prop-types";
import { useForm, FormContext } from "react-hook-form";

import { handleInputMask } from "helpers/masks";

import { CustomColumns } from "components/core/Grid";
import Button from "components/core/Button";
import Select from "components/core/Select";
import Loading from "components/core/Loading";

import * as S from "./styles";

const Filter = ({ isLoading, options, handleFilter }) => {
  const { register, handleSubmit, ...methods } = useForm({
    mode: "onChange",
    reValidateMode: "onBlur",
    submitFocusError: true,
  });

  const { setValue } = methods;

  return (
    <Loading isLoading={isLoading}>
      <S.Div>
        <FormContext setValue={setValue} {...methods}>
          <form autoComplete="off" onSubmit={handleSubmit(handleFilter)}>
            <CustomColumns minmax="150px">
              {options.map((option) => (
                <S.Wrapper key={option.id}>
                  {!option.values ? (
                    <S.Field
                      name={option.id}
                      id={option.id}
                      type="text"
                      ref={register}
                      onInput={(e) => {
                        option?.validation?.primitiveType &&
                          handleInputMask(
                            e,
                            option.validation.primitiveType,
                            option.validation?.entityType
                          );
                      }}
                    />
                  ) : (
                    <Select
                      register={register}
                      optionId={option.id}
                      values={option.values}
                    />
                  )}
                  <S.Label htmlFor={option.id}>{option.name}</S.Label>

                  {option.validation?.pattern && (
                    <S.Format>Formato: {option.validation?.pattern}</S.Format>
                  )}
                </S.Wrapper>
              ))}
            </CustomColumns>

            <Button margin="1rem 0 0">Filtrar</Button>
          </form>
        </FormContext>
      </S.Div>
    </Loading>
  );
};

Filter.propTypes = {
  isLoading: PropTypes.bool,
  handleFilter: PropTypes.func.isRequired,
  options: PropTypes.array,
};

Filter.defaultProps = {
  isLoading: false,
  options: [],
};

export default Filter;
