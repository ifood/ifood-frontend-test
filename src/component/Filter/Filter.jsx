import React, { useState, useMemo, useCallback } from 'react';
import PropTypes from 'prop-types';
import { Icon, Button } from 'semantic-ui-react';
import { useIntl } from 'react-intl';
import FilterModal from '../FilterModal';
import TagList from '../TagList';
import { Search, InputSearch, ModalTriggerWrapper } from './Filter.styled';
import { getFieldFormatter } from '../DynamicField/FieldMetadata';

function Filter({ metadata, filter, onFilter, onTextChange }) {
  const [modalOpen, setModalOpen] = useState(false);
  const intl = useIntl();

  function handleFilter(newFilter) {
    if (onFilter != null) {
      onFilter(newFilter);
    }
  }

  function handleDeleteTag(field) {
    const newFilter = { ...filter };
    delete newFilter[field];
    handleFilter(newFilter);
  }

  function handleTextChange(e) {
    if (onTextChange != null) {
      onTextChange(e?.target?.value);
    }
  }

  const tagFormatter = useCallback(
    ({ id, field, value }) => {
      const fieldMetadata = metadata?.find((item) => item?.id === field);
      if (fieldMetadata == null) {
        return { id, field, value };
      }
      let formatedValue = value;
      const formatter = getFieldFormatter(fieldMetadata);
      if (formatter != null) {
        formatedValue = formatter({ fieldMetadata, value });
      }
      return { id, field: fieldMetadata.name, value: formatedValue };
    },
    [metadata]
  );

  function handleOpenModal() {
    setModalOpen(true);
  }

  function handleCloseModal() {
    setModalOpen(false);
  }

  const tags = useMemo(
    function convertFilterObjectToList() {
      return filter == null
        ? null
        : Object.keys(filter).map((key) => {
            return tagFormatter({
              id: key,
              field: key,
              value: filter[key],
            });
          });
    },
    [filter, tagFormatter]
  );

  return (
    <>
      <Search>
        <InputSearch icon placeholder={intl.formatMessage({ id: 'search' })}>
          <input onChange={handleTextChange} />
          <Icon name="search" />
        </InputSearch>
        {metadata != null && (
          <ModalTriggerWrapper>
            <Button circular basic icon="filter" onClick={handleOpenModal} />
            <FilterModal
              filter={filter}
              metadata={metadata}
              open={modalOpen}
              onSubmit={handleFilter}
              onClose={handleCloseModal}
            />
          </ModalTriggerWrapper>
        )}
      </Search>
      <TagList
        title={intl.formatMessage({ id: 'filters' })}
        items={tags}
        onDeleteTag={handleDeleteTag}
        formatter={tagFormatter}
      />
    </>
  );
}

Filter.propTypes = {
  metadata: PropTypes.array,
  filter: PropTypes.object,
  onFilter: PropTypes.func,
  onTextChange: PropTypes.func,
};

export default React.memo(Filter);
