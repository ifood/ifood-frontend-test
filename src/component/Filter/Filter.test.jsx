import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';
import { Button } from 'semantic-ui-react';
import Filter from './Filter';
import FilterModal from '../FilterModal';
import TagList from '../TagList';

const metadata = [
  {
    id: 'number',
    name: 'Number',
    validation: {
      primitiveType: 'INTEGER',
    },
  },
];

const filterData = {
  number: '1',
  age: '50',
};

const formatedTag = {
  id: 'number', // filter field
  field: 'Number', // field metadata name
  value: '1', // field value
};

const OnChangeEvent = {
  target: {
    value: '',
  },
};

describe('<Filter />', () => {
  it('snapshot', () => {
    const filter = mount(<Filter metadata={metadata} filter={filterData} />);
    jestExpect(toJson(filter)).toMatchSnapshot();
  });

  it('should not render a modal or button if metadata is not provided', () => {
    const filter = shallow(<Filter />);
    expect(filter.find(Button)).to.have.lengthOf(0);
    expect(filter.find(FilterModal)).to.have.lengthOf(0);
  });

  describe('modal controller', () => {
    const filter = mount(<Filter metadata={metadata} />);

    it('should open the modal when button is clicked', () => {
      filter.find(Button).simulate('click');
      filter.update();
      expect(filter.find(FilterModal).props().open).to.be.true;
    });

    it('should close the modal when onClose callback is called', () => {
      act(() => {
        filter.find(FilterModal).props().onClose();
      });
      filter.update();
      expect(filter.find(FilterModal).props().open).to.be.false;
    });
  });

  describe('onTextChange callback', () => {
    it('should be called when search input is changed', () => {
      const onTextChange = jest.fn();
      const filter = shallow(<Filter onTextChange={onTextChange} />);
      filter.find('input').simulate('change', OnChangeEvent);
      expect(onTextChange.mock.calls).to.have.lengthOf(1);
    });

    it('should not throw an error if callback is not provided', () => {
      const filter = shallow(<Filter />);
      function change() {
        filter.find('input').simulate('change', OnChangeEvent);
      }
      expect(change).to.not.throw();
    });
  });

  describe('onFilter callback', () => {
    it('should be called when onSubmit callback is called', () => {
      const onFilter = jest.fn();
      const filter = shallow(
        <Filter metadata={metadata} onFilter={onFilter} />
      );
      filter.find(FilterModal).props().onSubmit();
      expect(onFilter.mock.calls).to.have.lengthOf(1);
    });

    it('should not throw an error if callback is not provided', () => {
      const filter = shallow(<Filter metadata={metadata} />);
      function submit() {
        filter.find(FilterModal).props().onSubmit();
      }
      expect(submit).to.not.throw();
    });

    it('should be called when a Tag is deleted', () => {
      const onFilter = jest.fn();
      const filter = shallow(
        <Filter metadata={metadata} onFilter={onFilter} />
      );
      filter.find(TagList).props().onDeleteTag();
      expect(onFilter.mock.calls).to.have.lengthOf(1);
    });
  });

  it('should format Tags', () => {
    const filter = shallow(<Filter metadata={metadata} filter={filterData} />);
    const tagList = filter.find(TagList);
    const tag = tagList.props().items[0];
    expect(tag).to.deep.equal(formatedTag);
  });
});
