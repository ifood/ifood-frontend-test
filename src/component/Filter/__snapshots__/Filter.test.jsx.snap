// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> snapshot 1`] = `
.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 5px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c2 {
  width: 55px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<Memo(Filter)
  filter={
    Object {
      "age": "50",
      "number": "1",
    }
  }
  metadata={
    Array [
      Object {
        "id": "number",
        "name": "Number",
        "validation": Object {
          "primitiveType": "INTEGER",
        },
      },
    ]
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <Styled(Input)
        icon={true}
        placeholder="intl::search"
      >
        <Input
          className="c1"
          icon={true}
          placeholder="intl::search"
          type="text"
        >
          <div
            className="ui icon input c1"
          >
            <input
              key=".0"
              onChange={[Function]}
              placeholder="intl::search"
              type="text"
            />
            <Icon
              as="i"
              key=".1"
              name="search"
            >
              <i
                aria-hidden="true"
                className="search icon"
                onClick={[Function]}
              />
            </Icon>
          </div>
        </Input>
      </Styled(Input)>
      <styled.div>
        <div
          className="c2"
        >
          <Button
            as="button"
            basic={true}
            circular={true}
            icon="filter"
            onClick={[Function]}
          >
            <Ref
              innerRef={
                Object {
                  "current": <button
                    class="ui basic circular icon button"
                  >
                    <i
                      aria-hidden="true"
                      class="filter icon"
                    />
                  </button>,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <button
                      class="ui basic circular icon button"
                    >
                      <i
                        aria-hidden="true"
                        class="filter icon"
                      />
                    </button>,
                  }
                }
              >
                <button
                  className="ui basic circular icon button"
                  onClick={[Function]}
                >
                  <Icon
                    as="i"
                    name="filter"
                  >
                    <i
                      aria-hidden="true"
                      className="filter icon"
                      onClick={[Function]}
                    />
                  </Icon>
                </button>
              </RefFindNode>
            </Ref>
          </Button>
          <FilterModal
            filter={
              Object {
                "age": "50",
                "number": "1",
              }
            }
            metadata={
              Array [
                Object {
                  "id": "number",
                  "name": "Number",
                  "validation": Object {
                    "primitiveType": "INTEGER",
                  },
                },
              ]
            }
            onClose={[Function]}
            onSubmit={[Function]}
            open={false}
          >
            <Modal
              centered={false}
              closeOnDimmerClick={true}
              closeOnDocumentClick={false}
              dimmer={true}
              eventPool="Modal"
              onClose={[Function]}
              open={false}
              size="tiny"
            >
              <Portal
                closeOnDocumentClick={false}
                closeOnEscape={true}
                eventPool="Modal"
                mountNode={<body />}
                onClose={[Function]}
                onMount={[Function]}
                onOpen={[Function]}
                onUnmount={[Function]}
                open={false}
                openOnTriggerClick={true}
              />
            </Modal>
          </FilterModal>
        </div>
      </styled.div>
    </div>
  </styled.div>
  <TagList
    formatter={[Function]}
    items={
      Array [
        Object {
          "field": "Number",
          "id": "number",
          "value": "1",
        },
        Object {
          "field": "age",
          "id": "age",
          "value": "50",
        },
      ]
    }
    onDeleteTag={[Function]}
    title="intl::filters"
  >
    <span>
      intl::filters
      :
    </span>
    <styled.div>
      <div
        className="c3"
      >
        <Tag
          field="Number"
          id="number"
          key="number"
          onDelete={[Function]}
          value="1"
        >
          <styled.div>
            <div
              className="c4"
            >
              <Button
                as="div"
                labelPosition="left"
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <div
                        class="ui left labeled button"
                        role="button"
                        tabindex="0"
                      >
                        <div
                          class="ui basic label"
                        >
                          Number
                          : 
                          1
                        </div>
                        <button
                          class="ui icon button"
                        >
                          <i
                            aria-hidden="true"
                            class="remove icon"
                          />
                        </button>
                      </div>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <div
                          class="ui left labeled button"
                          role="button"
                          tabindex="0"
                        >
                          <div
                            class="ui basic label"
                          >
                            Number
                            : 
                            1
                          </div>
                          <button
                            class="ui icon button"
                          >
                            <i
                              aria-hidden="true"
                              class="remove icon"
                            />
                          </button>
                        </div>,
                      }
                    }
                  >
                    <div
                      className="ui left labeled button"
                      onClick={[Function]}
                      role="button"
                      tabIndex={0}
                    >
                      <Label
                        basic={true}
                      >
                        <div
                          className="ui basic label"
                          onClick={[Function]}
                        >
                          Number
                          : 
                          1
                        </div>
                      </Label>
                      <Button
                        as="button"
                        icon={true}
                        onClick={[Function]}
                      >
                        <Ref
                          innerRef={
                            Object {
                              "current": <button
                                class="ui icon button"
                              >
                                <i
                                  aria-hidden="true"
                                  class="remove icon"
                                />
                              </button>,
                            }
                          }
                        >
                          <RefFindNode
                            innerRef={
                              Object {
                                "current": <button
                                  class="ui icon button"
                                >
                                  <i
                                    aria-hidden="true"
                                    class="remove icon"
                                  />
                                </button>,
                              }
                            }
                          >
                            <button
                              className="ui icon button"
                              onClick={[Function]}
                            >
                              <Icon
                                as="i"
                                name="remove"
                              >
                                <i
                                  aria-hidden="true"
                                  className="remove icon"
                                  onClick={[Function]}
                                />
                              </Icon>
                            </button>
                          </RefFindNode>
                        </Ref>
                      </Button>
                    </div>
                  </RefFindNode>
                </Ref>
              </Button>
            </div>
          </styled.div>
        </Tag>
        <Tag
          field="age"
          id="age"
          key="age"
          onDelete={[Function]}
          value="50"
        >
          <styled.div>
            <div
              className="c4"
            >
              <Button
                as="div"
                labelPosition="left"
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <div
                        class="ui left labeled button"
                        role="button"
                        tabindex="0"
                      >
                        <div
                          class="ui basic label"
                        >
                          age
                          : 
                          50
                        </div>
                        <button
                          class="ui icon button"
                        >
                          <i
                            aria-hidden="true"
                            class="remove icon"
                          />
                        </button>
                      </div>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <div
                          class="ui left labeled button"
                          role="button"
                          tabindex="0"
                        >
                          <div
                            class="ui basic label"
                          >
                            age
                            : 
                            50
                          </div>
                          <button
                            class="ui icon button"
                          >
                            <i
                              aria-hidden="true"
                              class="remove icon"
                            />
                          </button>
                        </div>,
                      }
                    }
                  >
                    <div
                      className="ui left labeled button"
                      onClick={[Function]}
                      role="button"
                      tabIndex={0}
                    >
                      <Label
                        basic={true}
                      >
                        <div
                          className="ui basic label"
                          onClick={[Function]}
                        >
                          age
                          : 
                          50
                        </div>
                      </Label>
                      <Button
                        as="button"
                        icon={true}
                        onClick={[Function]}
                      >
                        <Ref
                          innerRef={
                            Object {
                              "current": <button
                                class="ui icon button"
                              >
                                <i
                                  aria-hidden="true"
                                  class="remove icon"
                                />
                              </button>,
                            }
                          }
                        >
                          <RefFindNode
                            innerRef={
                              Object {
                                "current": <button
                                  class="ui icon button"
                                >
                                  <i
                                    aria-hidden="true"
                                    class="remove icon"
                                  />
                                </button>,
                              }
                            }
                          >
                            <button
                              className="ui icon button"
                              onClick={[Function]}
                            >
                              <Icon
                                as="i"
                                name="remove"
                              >
                                <i
                                  aria-hidden="true"
                                  className="remove icon"
                                  onClick={[Function]}
                                />
                              </Icon>
                            </button>
                          </RefFindNode>
                        </Ref>
                      </Button>
                    </div>
                  </RefFindNode>
                </Ref>
              </Button>
            </div>
          </styled.div>
        </Tag>
      </div>
    </styled.div>
  </TagList>
</Memo(Filter)>
`;
